File: rd-kt/rd-core/src/commonMain/kotlin/com/jetbrains/rd/util/threading/Linearization.kt
package com.jetbrains.rd.util.threading

import com.jetbrains.rd.util.Sync

// Puts linearization points into program starting with 0.
class Linearization {

    private var nextId = 0L
    private val lock = Any()


    private var enabled = true

    fun enable() {
        Sync.lock(lock) {
            enabled = true
            Sync.notifyAll(lock)
        }
    }

    fun disable() {
        Sync.lock(lock) {
            enabled = false
            Sync.notifyAll(lock)
        }
    }


    fun point(id: Int) {
        require(id >= 0) {"$id >= 0"}

        Sync.lock(lock) {
            while (enabled && id > nextId)
                Sync.wait(lock, 1000L)

            if (!enabled) return

            require (id <= nextId) {"Point $id already set, nextId=$nextId"}
            nextId ++
            Sync.notifyAll(lock)
        }
    }

    fun reset() {
        Sync.lock(lock) {
            nextId = 0L
            Sync.notifyAll(lock)
        }
    }
}

File: rd-net/Test.Lifetimes/Linearization.cs
using System.Threading;
using JetBrains.Diagnostics;

namespace Test.Lifetimes
{
  public class Linearization
  {
    private readonly object myLock = new object();
    
    private int myNextId;

    //could be disabled for some reasons
    private bool myEnabled;

    public void Enable()
    {
      lock (myLock)
      {
        myEnabled = true;
        Monitor.PulseAll(myLock);
      }
    }

    public void Disable()
    {
      lock (myLock)
      {
        myEnabled = false;
        Monitor.PulseAll(myLock);
      }
    }

    public void Point(int id)
    {
      Assertion.Require(id >= 0, "{0} >= 0", id);

      lock (myLock)
      {
        while (myEnabled && id > myNextId)
          Monitor.Wait(myLock, 1000);

        //could break waiting
        if (!myEnabled) return;
        
        Assertion.Require(id <= myNextId, "Point {0} already set, nextId={1}", id, myNextId);
        myNextId++; 
        
        Monitor.PulseAll(myLock);
      }
    }

    public void Reset()
    {
      lock (myLock)
      {
        myNextId = 0;
        Monitor.PulseAll(myLock);
      }
    }
  }
}